# Centromere Detection Pipeline

This pipeline performs centromere detection and scoring for fungal genomes using Nanopore and PacBio sequencing data. It is implemented as a Snakemake workflow designed for execution on an HPC system with optional GPU acceleration.

The pipeline automatically selects the appropriate processing path based on sequencing platform (Nanopore or PacBio).

## Required Software

The following software libraries were used in the creation of this pipeline. Unless otherwise stated, these were loaded using the installed versions available through the Tennessee Tech University HPC Spack v0.21.1 system installation. We've listed the versions that we used but have not extensively tested on downstream or upstream versions.

### Workflow and Environment Management
- snakemake@7.22.0
- singularityce@3.11.3 
- miniconda3@22.11.1
- graphviz@8.0.5 - not strictly necessary but useful for creating the visual DAG for the processing of the sample through the pipeline

### Core Bioinformatics Tools
- trf@4.09.1
- samtools@1.16.1 
- cdhit@4.8.1
- gffread@0.12.7
- bedops@2.4.41 
- minimap2@2.26^python@3.10 - The python version was added here to avoid conflicts with other dependencies on our system and, again, may not be strictly necessary.

The following were used to accelerate the use of the ccsmeth and pbccs libraries via a GPU. 
- py-biopython@1.81 
- py-torch@2.1.0 - spack hash /rvl 
- py-pybedtools@0.9.0
- py-scikit-learn@1.3.2
- py-statsmodels@0.14.0  

In addition to these, [ccsmeth](https://github.com/PengNi/ccsmeth) and [pbccs](https://ccs.how/) are necessary. They are available via conda install, however this pipeline makes the effort to create and implement the necessary conda environment (which may be found in envs/ccsmeth.yaml) therefore, manual installation of these softwares should not be necessary. 

The software [modbam2bed](https://github.com/epi2me-labs/modbam2bed) is not currently available via Spack on our system and was therefore installed from source using the following instructions:
```
git clone --recursive https://github.com/epi2me-labs/modbam2bed.git
make modbam2bed
./modbam2bed
```

The final software requirement is for the [edta](https://github.com/oushujun/EDTA) singularity container. This was installed using the [singularity instructions](https://github.com/oushujun/EDTA#install-with-singularity-good-for-hpc-users), specifically:
```
singularity build edta.sif docker://quay.io/biocontainers/edta:2.2.2--hdfd78af_1
```
## Setting up the file structure

The file structure for storing the original data files is as follows:
```
└── data
   └── nanopore
       └── Sample1
           ├── Sample1.fasta  
           ├── Sample1.fastq
           └── Sample1.gff3
       └── Sample2
           ├── Sample2.fasta  
           ├── Sample2.fastq
           └── Sample2.gff3
   └── pacbio
       └── Sample3
           ├── Sample3.fasta 
           ├── Sample3.gff3
           └── Sample3.subreads.bam   
```

For example:
```
└── data
   └── nanopore
       └── Guy11
           ├── Guy11.fasta  
           ├── Guy11.fastq
           └── Guy11.gff3
       └── Guy11_chr1
           ├── Guy11_chr1.fasta  
           ├── Guy11_chr1.fastq
           └── Guy11_chr1.gff3
   └── pacbio
       └── Fo4287v4
           ├── Fo4287v4.fasta 
           ├── Fo4287v4.gff3
           └── Fo4287v4.subreads.bam   
```

It is important to note that the Snakefile is looking for the samples to be in either a nanopore or pacbio directory then inside of a directory of the sample name and files must have the specific file extensions listed above.

The original data files must be set up in this manner and living in the correct nanopore or pacbio directory for the Snakefile to determine the proper set of steps to traverse for the pipeline. 

## About the config.yaml
The `config.yaml` will be your opportunity to customize the Snakefile to a certain extent. 

To run a sample you will need to add it to the config file in the following manner:

```
samples:
    nanopore:
        nanopore_sample_name:
    pacbio:
        pacbio_sample_name:
```

For example:
```
samples:
    nanopore:
        Guy11:
        Guy11_chr1:
    pacbio:
        Fo4287v4:
```

Additional parameters that you may wish to change:
- cpus_per_task - although we found that on our system performance was not improved beyond 12
- exclusion_bp_large #### FIXME: this needs an explanation
- exclusion_bp_min #### FIXME: this needs an explanation
- window – window size (in bp) used during centromere scoring
- trf #### FIXME: this needs an explanation
- te #### FIXME: this needs an explanation
- gene #### FIXME: this needs an explanation
- meth #### FIXME: this needs an explanation
- cov #### FIXME: this needs an explanation
- gc #### FIXME: this needs an explanation

The following paths must be updated to match your local environment:
```
# Singularity + EDTA
container:
  binds:
    - /work # This may be able to be removed altogether.
    - $(spack location -i ncbi-rmblastn):/rmblast # location of rmblast bound to container

edta:
  sif: edta.sif # Location of your edta.sif

# Meth Nanopore
modbam2bed: "modbam2bed/modbam2bed" # Location of modbam2bed executable

# Meth Pacbio
ccsmeth:
    call_mod:
        model_file: "../models/ # Location of your model files 
        model_ccsmeth_5mCpG_call_mods_attbigru2s_b21.v3.ckpt" # Specific model file to be used.

    call_freqb:
        model_file: "../models/model_ccsmeth_5mCpG_aggregate_attbigru_b11.v2p.ckpt"
```

## Running the Example Data
The Snakefile is looking for the specific output that you wish to create. 

We will use the files that have been used as an example throughout the documentation. 

1. Ensure that the needed software has been installed and loaded.
2. Add the data. For this example we will use the Guy11_chr1 sample data which will run through the nanopore portion of the pipeline. Those files should be placed in the CentromerDetection directory in data/nanopore/Guy11_chr1.
3. Set your config values according to the previous instructions.

You may run the entire pipeline with the following:

```
snakemake --use-conda --conda-frontend conda --cores 12 results/Guy11_chr1/CENTROMERE_SCORING/Guy11_chr1_1000/centro_candidates.bed

# Or to run all uncompleted steps on all samples:
snakemake --use-conda --conda-frontend conda --cores 12
```

Specifying the output file as the target will cause Snakemake to execute all required upstream steps automatically.